### PR TITLE
support async in agents

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

1. Support async methods in agent communication.
2. Add a notebook example for working with data stream.

Big thanks to @qingyun-wu for her previous hard work and exploration in this idea. Also thanks to @gagb @ekzhu @skzhang1 for their inspiration.

## Related issue number

This PR depends on #1177 
Related to: #1063 #1098 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/autogen/agentchat/agent.py [hunk 1]
```diff
@@ -46,3 +53,18 @@ def generate_reply(
         Returns:
             str or dict or None: the generated reply. If None, no reply is generated.
         """
+
+    def a_generate_reply(
```
Comment (qingyun-wu) [inline flaml/autogen/agentchat/agent.py:57]:
```suggestion
    async  def a_generate_reply(
```

**Review by ekzhu — CHANGES_REQUESTED** [review]

File: test/autogen/agentchat/test_async.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                result.clear()
+                return (
+                    True,
+                    f"Just got some latest market news. Merge your new suggestion with previous ones.\n{news_str}",
+                )
+            return False, None
+
+    user_proxy.register_auto_reply(autogen.AssistantAgent, add_data_reply, 1, context={"data": data})
+
+    await user_proxy.a_initiate_chat(
```
[the remark this answers is not in the document]
Comment (sonichi) [reply test/autogen/agentchat/test_async.py:103]: It'll be interesting. But I don't have an example now. Can we do that in a different PR?

File: test/autogen/agentchat/test_async.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    return feeds_summary
+
+
+async def test_stream():
+    try:
+        import openai
+    except ImportError:
+        return
+    config_list = autogen.config_list_from_json(OAI_CONFIG_LIST, KEY_LOC)
+    data = asyncio.Future()
```
[the remark this answers is not in the document]
Comment (sonichi) [reply test/autogen/agentchat/test_async.py:53]: The link you gave for asyncio.Queue says: "This class is [not thread safe](https://docs.python.org/3/library/asyncio-dev.html#asyncio-multithreading)."

This is just a test and not part of the core library.

File: test/autogen/agentchat/test_async.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            if result:
+                news_str = "\n".join(result)
+                result.clear()
+                return (
+                    True,
+                    f"Just got some latest market news. Merge your new suggestion with previous ones.\n{news_str}",
+                )
+            return False, None
+
+    user_proxy.register_auto_reply(autogen.AssistantAgent, add_data_reply, 1, context={"data": data})
```
[the remark this answers is not in the document]
Comment (sonichi) [reply test/autogen/agentchat/test_async.py]: renamed.

File: test/autogen/agentchat/test_async.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekzhu) [reply test/autogen/agentchat/test_async.py:103]: Yep.

**Review by ekzhu — APPROVED** [review]

**Comment by ekzhu** [comment]
As the next step after this PR, maybe we can think about the design of the message stream to manage async agent tasks.

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**